core-runner/src/runner.rs

```diff
@@ -313,7 +305,14 @@ impl<SC: StorageCarrier, E: Environment<Ext>> Runner<SC, E> {
             self.block_info,
         );
 
+        let outgoing_messages = context.message_buf.drain(..).collect::<Vec<_>>();
+
         if run_result.outcome.was_trap() && generate_reply_on_trap {
+
+            let gas_spent_for_outgoing: u64 = outgoing_messages.iter().map(|msg| msg.gas_limit).sum();
+            let burned_gas = run_result.gas_spent;
+            let trap_gas = incoming_message.gas_limit() - gas_spent_for_outgoing - burned_gas;
```

Comment:
How accurate the gas counter is? Will it hurt if we use saturating subtraction here (just in case)?